### PR TITLE
Custom Document provider

### PR DESCRIPTION
**Pull request by IvanIvanoff**
#### Summary
https://favro.com/organization/bdbb4f24afc48b29c74f4fa4/cffeb6e3eaf254588cb68a9a?card=San-4585

Short:
- Added a custom Absinthe DocumentProvider that excludes `Resolution` and `Result` from the cache if  the value can be retrieved from the cache
- Rework the apikey authorization to make less DB calls

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: config/test.exs [hunk 1]
```diff
@@ -21,13 +21,13 @@ config :tesla, adapter: Tesla.Mock
 config :sanbase, Sanbase.Repo,
   pool: Ecto.Adapters.SQL.Sandbox,
   database: "sanbase_test",
-  pool_size: 30
+  pool_size: 5
```
Comment (kalkov) [inline config/test.exs]: Why the pool size is lowered?

File: test/sanbase_web/graphql/context_plug_test.exs [hunk 2]
```diff
@@ -10,16 +10,24 @@ defmodule SanbaseWeb.Graphql.ContextPlugTest do
...
-             context: %{
-               auth: %{auth_method: :user_token, current_user: user},
-               remote_ip: {127, 0, 0, 1}
-             }
+    conn_context = conn.private.absinthe.context
+
+    assert conn_context.auth == %{auth_method: :user_token, current_user: user}
+    assert conn_context.remote_ip == {127, 0, 0, 1}
+
+    assert conn_context.permissions == %{
```
Comment (tspenov) [inline test/sanbase_web/graphql/context_plug_test.exs]: ==  `User.full_permissions()` ?

File: lib/sanbase_web/router.ex [hunk 3]
```diff
@@ -40,17 +40,23 @@ defmodule SanbaseWeb.Router do
...
       json_codec: Jason,
       schema: SanbaseWeb.Graphql.Schema,
+      document_providers: [
+        SanbaseWeb.Graphql.DocumentProvider,
+        Absinthe.Plug.DocumentProvider.Default
+      ],
       analyze_complexity: true,
       max_complexity: 10000,
-      interface: :simple
+      # interface: :simple,
```
Comment (tspenov) [inline lib/sanbase_web/router.ex]: commented line

File: lib/sanbase_web/router.ex [hunk 4]
```diff
@@ -40,17 +40,23 @@ defmodule SanbaseWeb.Router do
...
+      log_level: :info,
+      before_send: {SanbaseWeb.Graphql.Absinthe, :before_send}
     )
 
     forward(
       "/graphiql",
       Absinthe.Plug.GraphiQL,
       json_codec: Jason,
       schema: SanbaseWeb.Graphql.Schema,
+      document_providers: [
```
Comment (tspenov) [inline lib/sanbase_web/router.ex]: Do we need those type of caching for `graphiql` tool ?

File: config/test.exs (hunk 1 above)
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply config/test.exs]: Almost all tests are not concurrent, so higher pool size does not have an effect.
Apart from that, I started getting `too many connections` errors with Postgres 11

Comment (IvanIvanoff) [reply config/test.exs]: Almost all tests are not concurrent, so higher pool size does not have an effect.
Apart from that, I started getting `too many connections` errors with Postgres 11

File: lib/sanbase_web/router.ex (hunk 4 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/router.ex]: We want the same behavior for both /graphql and /graphiql

File: lib/sanbase_web/graphql/document/utils.ex [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            {:ok, datetime, _} -> {key, datetime}
+            _ -> {key, value}
+          end
+
+        pair ->
+          pair
+      end)
+      |> Map.new()
+
+    # Cache for between 30 seconds and 90 seconds
```
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/document/utils.ex]: It's not out of sync - the cache is indeed for 30-90 seconds. Nevertheless, I'll remove it

File: lib/sanbase_web/graphql/absinthe.ex [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  def before_send(conn, %Absinthe.Blueprint{} = blueprint) do
+    queries =
+      blueprint.operations
+      |> Enum.flat_map(fn %{selections: selections} ->
+        selections
+        |> Enum.map(fn %{name: name} -> name end)
+      end)
+
+    all_queries_cachable? = Enum.all?(queries, &Enum.member?(@cached_queries, &1))
```
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/absinthe.ex]:
```
However: in draft -08 (and in RFC2068), "cacheable" appears only
once, and "cachable" appears dozens of times.  I vote to stick
with "cachable."
```
sticking with `cachable` :smiley:

File: lib/sanbase_web/graphql/document/document_provider.ex [hunk 7]
```diff
@@ -0,0 +1,85 @@
+defmodule SanbaseWeb.Graphql.DocumentProvider do
+  @moduledoc ~s"""
+  Custom Absinthe DocumentProvider for more effective caching.
+
+  The `ContextPlug` runs before this DocumentProvider. In this plug the context is
+  created. It checks the Authorization header and adds the user, permissins and cache key
```
Comment (tspenov) [inline lib/sanbase_web/graphql/document/document_provider.ex]: typo: `permissions`

File: lib/sanbase_web/graphql/document/document_provider.ex [hunk 8]
```diff
@@ -77,6 +77,7 @@ defmodule SanbseWeb.Graphql.Phase.Document.Execution.Resolution do
   def run(bp_root, _) do
     # Will be fetched from cache - that's what's saved by SanbaseWeb.Graphql.Absinthe.before_send
     result = Process.get(@cache_dict_key)
+    Process.put(:do_not_cache_query, true)
```
Comment (tspenov) [inline lib/sanbase_web/graphql/document/document_provider.ex]: 👍

**Pull request merged**